storage/store_pool.go

```diff
@@ -0,0 +1,212 @@
+// for names of contributors.
+//
+// Author: Bram Gruneir (bram+code@cockroachlabs.com)
+
+package storage
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/stop"
+)
+
+// TestTimeUntilStoreDead is the default TimeUntilStoreDead for running tests.
+const TestTimeUntilStoreDead = 10 * time.Millisecond
+
+type storeDetail struct {
+	desc            proto.StoreDescriptor
+	dead            bool
+	timesDied       int
+	foundDeadOn     time.Time
+	lastUpdatedTime time.Time // This is also the priority for the queue.
+	index           int       // index of the item in the heap, required for heap.Interface
+}
+
+// storePoolPQ implements the heap.Interface (which includes  sort.Interface)
+// and holds storeDetail.
+type storePoolPQ []*storeDetail
+
+// Len implements the sort.Interface.
+func (pq storePoolPQ) Len() int {
+	return len(pq)
+}
+
+// Less implements the sort.Interface.
+func (pq storePoolPQ) Less(i, j int) bool {
+	return pq[i].lastUpdatedTime.UnixNano() < pq[j].lastUpdatedTime.UnixNano()
+}
+
+// Swap implements the sort.Interface.
+func (pq storePoolPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index, pq[j].index = i, j
+}
+
+// Push implements the heap.Interface.
+func (pq *storePoolPQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*storeDetail)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop implements the heap.Interface.
+func (pq *storePoolPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// peek returns the next value in the priority queue without dequeuing it.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) peek() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := (*pq)[0]
+	return detail
+}
+
+// enqueue either adds the detail to the queue or updates its location in the
+// priority queue.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) enqueue(detail *storeDetail) {
+	if detail.index < 0 {
+		heap.Push(pq, detail)
+	} else {
+		heap.Fix(pq, detail.index)
+	}
+}
+
+// dequeue removes the next detail from the priority queue.
+// Lock is assumed help by the caller.
+func (pq *storePoolPQ) dequeue() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := heap.Pop(pq).(*storeDetail)
+	return detail
+}
+
+// StorePool maintains a list of all known stores in the cluster and
+// information on their health.
+type StorePool struct {
+	gossip             *gossip.Gossip
+	clock              *hlc.Clock
+	timeUntilStoreDead time.Duration
+
+	mu     sync.RWMutex // Protects variables below...
+	stores map[proto.StoreID]*storeDetail
+	queue  *storePoolPQ
```

Comment:
True, but it makes the rest of the code cleaner for all calls to enqueue/dequeue/peek.  I can change it if you want.